Respect IModelNameProvider when matching OpenAPI parameters

khellang opened the pull request:
## Description

This makes sure any attribute implementing `IModelNameProvider` is respected when matching OpenAPI operation parameters with .NET action parameters.

It also replaces the `CancellationToken` workaround from the initial review with a proper check against `BindingSource.Body`, so documentation for parameters that aren't part of the OpenAPI surface (e.g. `CancellationToken` or `[FromServices]` parameters) is no longer applied to the request body description.

Fixes #64534
Fixes #64521
Fixes #67076
Fixes #65805

khellang committed:
Respect IModelNameProvider when matching OpenAPI parameters

khellang committed:
Add test demonstrating the issue

khellang committed:
Extract helper method to get model names

khellang committed:
Also respect IModelNameProvider when getting XML docs for properties

khellang committed:
Fix nullability warnings from Copilot

khellang commented on src/OpenApi/gen/XmlCommentGenerator.Emitter.cs:
@captainsafia Do you know why this fallback exists? We're seeing the `CancellationToken` argument being applied to the request body because of this, which is weird 🤔

khellang committed:
Add test to verify cancellation token description

khellang committed:
Don't include CancellationToken when applying XML doc description

khellang committed:
Remove unused usings

khellang committed:
Make assert less specific to avoid regressions

desjoerd commented on src/OpenApi/test/Microsoft.AspNetCore.OpenApi.SourceGenerators.Tests/OperationTests.Controllers.cs:
```diff
@@ -124,7 +124,7 @@ public class TestController : ControllerBase
     /// <param name="userId">The id of the user.</param>
     [HttpGet("{userId}")]
-    public string Get()
+    public string Get(int userId)
```
I am fine if you rename it. Or maybe we can just add a comment in the description of that controller, to keep the diff a bit smaller in regards to snapshots

khellang committed:
Add test for Minimal APIs

khellang committed:
Don't fetch model names if there are no parameters

khellang committed:
Add back missing parameter and fix support for missing parameters

mikekistler reviewed (commented):
Looks good! 👍 

@captainsafia I will defer to you for formal approval, but this all looks good to me.

khellang commented:
Hello @dotnet/minimal-apis! 👋🏻 I would like to request a review on this PR. It's been sitting here since November, right after .NET 10 came out and these bugs were discovered. That's almost half a year. Since then it's been a blocker to adopt .NET 10 and currently there's no indication if/when this will be merged. I must say this is pretty poor handling of community contributions fixing real customer issues 😞  At the very least, can you give us a sign of life? Is there anyone on the other side of the ether?

BrennanConroy commented:
Sorry @khellang, we've had a bunch of changes to the team these past few months which has sadly affected valuable contributions like this. I've brought this up to our team to review so we'll hopefully be able to have PR feedback/merge soon.

It looks good from my first quick glance, but will either defer to others or will do a more thorough review in a couple days. Is the cancellation part just a nice addition you found while working on this?

BrennanConroy reviewed (approved):
Could you file a follow-up issue to look at other parameter types besides CancellationToken? e.g. HttpContext, [FromServices], ClaimsPrincipal, etc.

Might want to look at the ApiParameterDescription.BindingSource?

khellang commented:
Thanks @BrennanConroy! Filed #67076. Can probably take a stab at fixing it as well if no one beats me to it.

BrennanConroy commented:
/azp run

BrennanConroy commented:
The 2 new tests in src/OpenApi/test/Microsoft.AspNetCore.OpenApi.SourceGenerators.Tests/OperationTests.Controllers.cs failed. I think the verified file(s) need to be regenerated.

khellang committed:
Update stale snapshots after rebase

khellang committed:
Only apply parameter documentation to the request body when the parameter is bound to it

This avoids applying documentation for parameters that aren't part of
the OpenAPI surface, like CancellationToken or [FromServices] parameters,
to the request body description.

khellang commented:
Rebased and updated the stale snapshots. I also included a commit that checks the `BindingSource` instead of the hard-coded `CancellationToken` check. It should fix https://github.com/dotnet/aspnetcore/issues/67076 as well.

BrennanConroy commented on src/OpenApi/gen/XmlCommentGenerator.Emitter.cs:
```diff
@@ -389,7 +366,7 @@ public Task TransformAsync(OpenApiOperation operation, OpenApiOperationTransform
                     {
                         var parameterInfo = methodInfo.GetParameters().SingleOrDefault(info => info.Name == parameterComment.Name);
-                        var operationParameter = operation.Parameters?.SingleOrDefault(parameter => parameter.Name == parameterComment.Name);
+                        var operationParameter = GetOperationParameter(operation, parameterInfo, parameterComment);
```
Not sure we care too much, but if it's easy to fix could we avoid calling `GetModelNames` multiple times per parameter?

BrennanConroy commented on src/OpenApi/gen/XmlCommentGenerator.Emitter.cs:
```diff
@@ -499,6 +483,86 @@ public Task TransformAsync(OpenApiOperation operation, OpenApiOperationTransform
+
+            foreach (var parameterDescription in context.Description.ParameterDescriptions)
+            {
+                if (parameterDescription.Source == BindingSource.Body
```
This excludes `IFormFile` (and other form types) from the description. Should those be excluded?

khellang replied:
But those _are_ bound to the body, right?

BrennanConroy replied:
Right, so if there is a description on the form param it should end up as the description in openapi?
https://github.com/dotnet/aspnetcore/issues/67078 calls this out as well.

BrennanConroy commented on src/OpenApi/gen/XmlCommentGenerator.Emitter.cs:
```diff
@@ -449,10 +429,14 @@ public Task TransformAsync(OpenApiOperation operation, OpenApiOperationTransform
+                    if (XmlCommentCache.Cache.TryGetValue(DocumentationCommentIdHelper.NormalizeDocId(propertyInfo.CreateDocumentationId()), out var propertyComment))
                     {
-                        var parameter = operation.Parameters?.SingleOrDefault(p => p.Name == metadata.Name);
+                        var parameter = GetOperationParameter(operation, propertyInfo);
```
Do we need to apply `IsRequestBodyParameter` to this piece of code as well?

khellang committed:
Address review feedback on request-body parameter handling

- Compute model names once per parameter instead of recomputing in
  GetOperationParameter and IsRequestBodyParameter
- Treat form/form-file binding sources as request-body parameters so
  form parameter documentation is applied to the request body
- Guard the property-loop request-body fallback with the binding source
  so only body-bound properties apply their docs to the request body

khellang commented:
Rebased and fixed latest comments. Sorry for the long delay 😞

khellang committed:
Update stale snapshot for test added on main

Youssef1313 commented:
@khellang This fixes #65805 as well right?

khellang commented:
> @khellang This fixes #65805 as well right?

Indeed. The PR predates the issue itself, but added a mention in the description now 👍

Youssef1313 commented on src/OpenApi/test/Microsoft.AspNetCore.OpenApi.SourceGenerators.Tests/snapshots/AddOpenApiTests.CanInterceptAddOpenApi#OpenApiXmlCommentSupport.generated.verified.cs:
```diff
@@ -371,7 +348,8 @@ public Task TransformAsync(OpenApiOperation operation, OpenApiOperationTransform
                     {
                         var parameterInfo = methodInfo.GetParameters().SingleOrDefault(info => info.Name == parameterComment.Name);
-                        var operationParameter = operation.Parameters?.SingleOrDefault(parameter => parameter.Name == parameterComment.Name);
+                        var modelNames = GetModelNames(parameterInfo, parameterInfo?.Name ?? parameterComment.Name);
```
I think if `parameterInfo` is not null, we know for sure that `parameterInfo.Name` is the same as `parameterComment.Name`. So we can just use `parameterComment.Name`.

```suggestion
                        var modelNames = GetModelNames(parameterInfo, parameterComment.Name);
```

Youssef1313 commented on src/OpenApi/test/Microsoft.AspNetCore.OpenApi.SourceGenerators.Tests/snapshots/AddOpenApiTests.CanInterceptAddOpenApi#OpenApiXmlCommentSupport.generated.verified.cs:
```diff
@@ -431,10 +412,15 @@ public Task TransformAsync(OpenApiOperation operation, OpenApiOperationTransform
                 {
-                    var propertyDocId = DocumentationCommentIdHelper.CreateDocumentationId(containerType, propertyName);
-                    if (XmlCommentCache.Cache.TryGetValue(DocumentationCommentIdHelper.NormalizeDocId(propertyDocId), out var propertyComment))
+                    var propertyInfo = containerType.GetProperty(propertyName);
```
Should this use `BindingFlags.DeclaredOnly`?

Otherwise, a shadowing property will cause `GetProperty` to throw an ambiguous match exception.

khellang replied:
Huh, that's something I've never considered or encountered in my 17 years as a .NET developer. 😅 I guess just switching to `DeclaredOnly` isn't enough either, as that will break the even more common case of inherited properties. I'd have to walk the hierarchy upwards combined with `DeclaredOnly` to get both.

Youssef1313 replied:
I was under the impression that in this code path, `propertyName` is guaranteed to be exactly declared on `containerType` and not on a base type. But maybe I'm wrong. When we create the metadata, there are two paths where it's guaranteed that the `containerType` is the exact type declaring the property:

https://github.com/dotnet/dotnet/blob/4a6c691f97cd93b08ef0794d297de622e4409e66/src/aspnetcore/src/Mvc/Mvc.ApiExplorer/src/EndpointMetadataApiDescriptionProvider.cs#L528

https://github.com/dotnet/dotnet/blob/4a6c691f97cd93b08ef0794d297de622e4409e66/src/aspnetcore/src/Mvc/Mvc.Core/src/ModelBinding/Metadata/DefaultModelMetadataProvider.cs#L179

But one code path where that **might** not be the case is:

https://github.com/dotnet/dotnet/blob/4a6c691f97cd93b08ef0794d297de622e4409e66/src/aspnetcore/src/Mvc/Mvc.Core/src/ModelBinding/Metadata/DefaultModelMetadataProvider.cs#L342-L345

I can't say for sure without looking in more details. Having that covered with tests will be appreciated though.

Youssef1313 replied:
I think even for the third code path, it's guaranteed that the property name and container type won't involve inheritance.

https://github.com/dotnet/dotnet/blob/4a6c691f97cd93b08ef0794d297de622e4409e66/src/aspnetcore/src/Shared/PropertyHelper/PropertyHelper.cs#L445-L449

https://github.com/dotnet/dotnet/blob/4a6c691f97cd93b08ef0794d297de622e4409e66/src/aspnetcore/src/Shared/PropertyHelper/PropertyHelper.cs#L477

khellang replied:
You're right about `GetCacheEntry(PropertyInfo, modelType)` (`DefaultModelMetadataProvider.cs:179`), where `ContainerType == property.DeclaringType`. But the path ApiExplorer actually uses for expanding bound properties is `GetMetadataForProperties` → `CreatePropertyDetails` (`:335`), which builds `ForProperty(property, …, key.ModelType)` (`:342`) — so `ContainerType` is the *model type* being bound, which is the **derived** type when a derived model has an inherited property. `GetVisibleProperties` only collapses `new`-shadowed members; genuinely inherited members keep their base `DeclaringType`.

That means `DeclaredOnly` on its own would return `null` for an inherited property and silently drop its description - a regression on a pretty common case. The hierarchy walk keeps inherited properties working *and* avoids the `AmbiguousMatchException` on shadowed ones. I'll add a test that binds a derived model with an inherited, documented property so this is covered (it fails with `DeclaredOnly`-alone and passes with the walk).

Youssef1313 commented on src/OpenApi/test/Microsoft.AspNetCore.OpenApi.SourceGenerators.Tests/snapshots/AddOpenApiTests.CanInterceptAddOpenApi#OpenApiXmlCommentSupport.generated.verified.cs:
```diff
@@ -481,6 +468,77 @@ public Task TransformAsync(OpenApiOperation operation, OpenApiOperationTransform
+                return modelNames;
+            }
+
+            foreach (var attribute in attributeProvider.GetCustomAttributes(inherit: false))
```
Why false?

khellang committed:
Address review feedback: simplify name lookup, guard GetProperty, inherit attributes

- Simplify GetModelNames call to use parameterComment.Name directly
- Add GetPropertyInfo helper to avoid AmbiguousMatchException on shadowed properties while preserving inherited-property resolution
- Read model-name attributes with inherit: true

Youssef1313 commented on src/OpenApi/test/Microsoft.AspNetCore.OpenApi.SourceGenerators.Tests/snapshots/AdditionalTextsTests.CanHandleXmlForSchemasInAdditionalTexts#OpenApiXmlCommentSupport.generated.verified.cs:
```diff
@@ -510,6 +497,94 @@ public Task TransformAsync(OpenApiOperation operation, OpenApiOperationTransform
+        {
+            // Walk the type hierarchy explicitly with `DeclaredOnly` so that a property shadowing a
+            // base member (via `new`) doesn't throw an `AmbiguousMatchException`, while still
+            // resolving properties that are inherited from a base type.
```
I don't feel walking the inheritance hierarchy is needed.

If it's really needed, can we have a test for it?

khellang replied:
So throw `AmbiguousMatchException` or exclude inherited properties?

Youssef1313 replied:
I guess excluding should be fine.

So you use `DeclaredOnly` and not walk the inheritance hierarchy.

@BrennanConroy Thoughts please? See my previous comment for reference: https://github.com/dotnet/aspnetcore/pull/64535#discussion_r3547028578

khellang replied:
Added a test that fails with DeclaredOnly, but passes with the inheritance chain walk. That would clearly be a regression if we let it through.

khellang committed:
Add test for XML comments on inherited properties from controllers

Covers the case where a bound model inherits a documented property from a
base type. The property is enumerated under the derived ContainerType while
its DeclaringType is the base, so resolving it requires walking the hierarchy
(a lookup restricted to the container type alone would drop the description).

khellang committed:
Add test for XML comments on shadowed properties from controllers

Covers a bound model that redeclares a base property with `new` and a
different type. A plain Type.GetProperty(name) throws AmbiguousMatchException
for this shape; the property lookup must resolve the most-derived declaration
and apply its documentation.

Youssef1313 commented:
Hi @khellang 

I'd like to get this in .NET 11 Preview 7. Are you planning to complete the work here please? Let me know if there is something we can help with, or if you want us to take over the PR to completion.

khellang commented:
Great! I'm not aware of any remaining work?

Youssef1313 commented:
@khellang The build is failing with:

```
/Users/runner/work/1/s/artifacts/obj/Sample/Release/net11.0/generated/Microsoft.AspNetCore.OpenApi.SourceGenerators/Microsoft.AspNetCore.OpenApi.SourceGenerators.XmlCommentGenerator/OpenApiXmlCommentSupport.generated.cs(593,36): error IL2070: 'this' argument does not satisfy 'DynamicallyAccessedMemberTypes.PublicProperties' in call to 'System.Type.GetProperty(String, BindingFlags)'. The parameter 'containerType' of method 'Microsoft.AspNetCore.OpenApi.Generated.XmlCommentOperationTransformer.GetPropertyInfo(Type, String)' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it is assigned to. [/Users/runner/work/1/s/src/OpenApi/sample/Sample.csproj]
/Users/runner/work/1/s/artifacts/obj/Sample/Release/net11.0/generated/Microsoft.AspNetCore.OpenApi.SourceGenerators/Microsoft.AspNetCore.OpenApi.SourceGenerators.XmlCommentGenerator/OpenApiXmlCommentSupport.generated.cs(593,36): error IL2075: 'this' argument does not satisfy 'DynamicallyAccessedMemberTypes.PublicProperties' in call to 'System.Type.GetProperty(String, BindingFlags)'. The return value of method 'System.Type.BaseType.get' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it is assigned to. [/Users/runner/work/1/s/src/OpenApi/sample/Sample.csproj]
```

khellang commented:
OK, I'm on vacation and probably wont be able to get it fixed right away. Let me see if I can get Claude to fix it...

claude committed:
Suppress trimming warnings for property lookup in generated XML comment transformer

The generated XmlCommentOperationTransformer.GetPropertyInfo helper reflects
over the container type (from ModelMetadata.ContainerType) via Type.GetProperty
and a BaseType walk. That container type is not statically annotated for
trimming, so the emitted code produced IL2070/IL2075 analyzer errors when the
OpenApi Sample project is built with trim analysis enabled.

Annotating the parameter would only move the warning to the call site (the
source Type is runtime model metadata). Instead, suppress IL2070/IL2075 on the
helper: if a property has been trimmed, GetProperty returns null and the XML
documentation is simply not applied, so this degrades safely.

Applied identically to the source generator emitter template and all affected
Verify snapshots so the source-generator snapshot tests stay in sync.

khellang commented:
Better now?

Youssef1313 commented on src/OpenApi/test/Microsoft.AspNetCore.OpenApi.SourceGenerators.Tests/snapshots/AdditionalTextsTests.CanHandleXmlForSchemasInAdditionalTexts#OpenApiXmlCommentSupport.generated.verified.cs:
```diff
@@ -510,6 +497,96 @@ public Task TransformAsync(OpenApiOperation operation, OpenApiOperationTransform
+            return null;
+        }
+
+        private static IReadOnlySet<string> GetModelNames(ICustomAttributeProvider? attributeProvider, string? name)
```
Why do we need a set here? Should we only consider `IModelNameProvider` if present?

And instead of special casing `IModelNameProvider`, could we instead access the binding info from the context through `context.Description.ActionDescriptor.Parameters[...].BindingInfo.BinderModelName` and/or `context.Description.ActionDescriptor.Parameters[...].Name`?

khellang replied:
The current code fixes all the bugs. Feel free to make whatever changes you want.

Youssef1313 replied:
I removed the HashSet and added a test that captures a behavior difference here. Other than that, this LGTM. Thanks for the contribution @khellang

Youssef1313 committed:
Use only IModelNameProvider when exists and add regression test

Youssef1313 commented:
/backport to release/11.0-preview7
